gstreamer-1.16.0

nh2 opened the pull request:
* [x] Based upon the (tiny) #54380; let's merge that first. (**done**)
* [x] Wait for 1.16 to be released (https://github.com/NixOS/nixpkgs/pull/54398#issuecomment-455932989)
* [x] Fix any fallout when 1.16 is released

## Motivation for this change

~The latest gstreamer version~ The current gstreamer development version, leading up to the next to-be-released 1.16 stable version. Already doing some work here so we have less work to do when 1.16 comes out.

**Edit:** This is now, and the PR has been updated.

Copying an important part of the commit message:

---

During the 1.14 -> 1.15 upgrade, lots of stuff stopped working because gstreamer changed what features are enabled by default and which ones are automatically turned on/off via pkgconfig dependency detection.

This resulted in the `gstreamer` ("core" attribute in nixpkgs) package to have only 15 of its previous 163 build targets enabled, and downstream packages breaking correspondingly.

To ease maintainability and to ensure users will find the expected features available (and when not, will see in the nix file why not), we now pass the `-Dauto_features=enabled` Meson build flag to all gstreamer builds, which sets all `auto` dependencies to `enabled`, and we explicitly disable those that we can't build .

This means in particular that `gst-plugins-bad` now has vastly more integrations (namely all for which nixpkgs has libraries available).

## @nh2's notes while working on this (for posterity)

* various meson patches: Removed: Seem no longer necessary, but I haven't checked all the details and whether the patches were fully merged or only partially.
  See e.g. https://lists.freedesktop.org/archives/gstreamer-bugs/2018-August/223857.html where they say
  > Committed without the gl part which was installed into libdir on purpose. You will have to take account of that in your packaging.
* libice: `gstreamer-check-1.0.pc` not found (and indeed it's not there)
  * fixed by `-Dcheck=enabled` meson flag, see https://github.com/GStreamer/gstreamer/blob/6ea4380230d1cca1d63196dfbd753af887b8b241/pkgconfig/meson.build#L32
* `core` configure output:
  ```diff
  -Found pkg-config: /nix/store/4iij6zf2z7gx68rx3kg28xi5rqvnfja6-pkg-config-0.29.2/bin/pkg-config (0.29.2)
  -Dependency libunwind found: YES 1.3.1
  -Found CMake: NO
  -Dependency libdw found: NO (tried pkgconfig and cmake)
  -Message: Support for backtraces is partial only.
  +Dependency libunwind skipped: feature libunwind disabled
  +Dependency libdw skipped: feature libdw disabled
  +Checking for function "backtrace" : YES
  ```
  * likely the same as before (partial support), but I improved it by adding an `elfutils` dep which provides `libdw`
* `core` configure output:
  -Build targets in project: 163
  +Build targets in project: 15
* `gst-validate` build complains that `GstPbutils-1.0.gir` couldn't be found / `GEN`'d even though it says `checking for GST_PBUTILS... yes` before
  * probably due to `core` configure output: `Program g-ir-scanner skipped: feature introspection disabled`

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Assured whether relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Tested via:

`NIX_PATH=nixpkgs=. nix-shell -p gst_all_1.gstreamer -p gst_all_1.gst-plugins-good -p gst_all_1.gst-plugins-base -p gst_all_1.gst-plugins-bad -p gst_all_1.gst-plugins-ugly -p gst_all_1.gst-libav -p gst_all_1.gst-rtsp-server -p gst_all_1.gst-validate -p gst_all_1.gst-vaapi`

CC @matthewbauer

## TODO

* [x] Merge #54380
* [x] Somebody should double-check whether indeed all patches removed in this PR have all their contents merged upstream, or if some were merged only partially
  * Done in https://github.com/NixOS/nixpkgs/pull/54398#issuecomment-455939823
* [x] Check that the vastly-increased `gst-plugins-bad` integrations work on Darwin, make those conditional that don't
  * See #60922
* [ ] This needs some real testing from some real NixOS users, as I haven't checked whether any of the multimedia functionality actually works (I only built it)

nh2 committed:
gstreamer: 1.14.4 -> 1.15.1, and extended gst-plugins-bad.

During the 1.14 -> 1.15 upgrade, lots of stuff stopped working because
gstreamer changed what features are enabled by default and which ones are
automatically turned on/off via pkgconfig dependency detection.

This resulted in the `gstreamer` ("core" attribute in nixpkgs) package
to have only 15 of its previous 163 build targets enabled, and downstream
packages breaking correspondingly.

To ease maintainability and to ensure users will find the expected features
available (and when not, will see in the nix file why not), we now pass
the `-Dauto_features=enabled` Meson build flag to all gstreamer builds,
which sets all `auto` dependencies to `enabled`, and we explicitly disable
those that we can't build.

This means in particular that `gst-plugins-bad` now has vastly more integrations
(namely all for which nixpkgs has libraries available).

jtojnar commented:
I did not enable `gtksink` by default due to its closure size (150 MB vs 300 MB).

Also please note that 1.15 is an unstable development branch for 1.16 stable release.

nh2 commented:
> I did not enable `gtksink` by default due to its closure size (150 MB vs 300 MB).

@jtojnar That makes sense; the PR doesn't change this, it is still optional.

> Also please note that 1.15 is an unstable development branch for 1.16 stable release.

Oops OK. Let's wait until 1.16 before continuing with this then.

jtojnar commented:
The removed patches come from https://bugzilla.gnome.org/show_bug.cgi?id=794856 and have been committed.

nh2 commented:
> The removed patches come from https://bugzilla.gnome.org/show_bug.cgi?id=794856 and have been committed.

@jtojnar Thanks! I ticked off the bullet point in the issue description based on this feedback.

> closure size

We will also want to do a closure size check on the added `-bad` dependencies eventually.

I think it makes sense that it'll increase and will likely be worth it, as for end users nothing is more annoying than a gstreamer plugin but it not working on your machine because the distro hasn't enabled it, and `-bad` is supposed to contain as many integrations as possible, but we should still know the number.

Also, I wouldn't be opposed to introducing feature flags as arguments to the `callPackage` for each of the external dependencies so that users can easily build a smaller `-bad` package for their use cases, but I think the default should be `? true` for most features so that in the default case multimedia "just works".

jtojnar commented:
115.2 was released, 1.16 in few weeks: https://gstreamer.freedesktop.org/news/#2019-02-27T10:00:00Z

nh2 committed:
gstreamer: 1.15.1 -> 1.16.0

Changes needed:

* core: dbghelp disabled as it's not needed
* core: Fixed script not being executable and skipped by patchShebangs
* base: gtk_doc disabled because of wayland symbol error
* ges: Patch removed that was merged upstream:
  https://bugzilla.gnome.org/show_bug.cgi?id=794856#c16

nh2 commented:
1.16 was released.

I've updated the PR with another commit to update to 1.16 and fix a few build errors.

I've also made an upstream PR to fix a script not being executable, which broke the build:

https://gitlab.freedesktop.org/gstreamer/gstreamer/merge_requests/141

nh2 commented:
Let's see if this works:

@grahamcofborg build gst_all_1

adisbladis commented:
@GrahamcOfBorg build gst_all_1

jtojnar commented on pkgs/development/libraries/gstreamer/bad/default.nix:
```diff
@@ -85,6 +137,38 @@ stdenv.mkDerivation rec {
+    "-Dopenmpt=disabled" # `libopenmpt` not packaged in nixpkgs as of writing
+    "-Dopenni2=disabled" # not packaged in nixpkgs as of writing
+    "-Dopensles=disabled" # not packaged in nixpkgs as of writing
+    "-Drtmp=disabled" # `librtmp` not packaged in nixpkgs as of writing
```
This is in `rtmpdump`.

nh2 replied:
Done.

jtojnar commented on pkgs/development/libraries/gstreamer/bad/default.nix:
```diff
@@ -85,6 +137,38 @@ stdenv.mkDerivation rec {
+
+    "-Ddts=disabled" # required `libdca` library not packaged in nixpkgs as of writing, and marked as "BIG FAT WARNING: libdca is still in early development"
+    "-Dfaac=${if faacSupport then "enabled" else "disabled"}"
+    "-Dgsm=disabled" # as of writing, with `gsm` in `buildInputs` we get "GSM plugin is enabled: found libgsm but no headers"; gsm packaging problem?
```
Looks like a typo, there is also incorrect indentation: https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/blob/3a0a82a783b2a8af4c0962bcf3b238ca927e4714/ext/gsm/meson.build#L8

nh2 replied:
Yes looks like it.

Even when fixing it (see my [upstream PR](https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/merge_requests/312) I made based on your comment), it still fails to compile because of:

https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/merge_requests/312#note_155041

nh2 replied:
Fixed that too now: https://gitlab.freedesktop.org/gstreamer/gst-plugins-bad/merge_requests/312#note_155052

nh2 replied:
Done

jtojnar commented on pkgs/development/libraries/gstreamer/bad/default.nix:
```diff
@@ -85,6 +137,38 @@ stdenv.mkDerivation rec {
+    "-Dgsm=disabled" # as of writing, with `gsm` in `buildInputs` we get "GSM plugin is enabled: found libgsm but no headers"; gsm packaging problem?
+    "-Diqa=disabled" # required `dssim` library not packaging in nixpkgs as of writing
+    "-Dmsdk=disabled" # not packaged in nixpkgs as of writing / no Windows support
+    "-Dmusepack=disabled" # as of writing, with `libmpcdec` in `buildInputs` we get "Could not find libmpcdec header files, but Musepack was enabled via options"; libmpcdec packaging problem?
```
We have the header in `libmpc/mpcdec.h` instead of `mpc/mpcdec.h` like [Arch does](https://www.archlinux.org/packages/extra/x86_64/libmpcdec/). Not sure who is at fault.

nh2 replied:
The situation looks difficult. There are apparently 2 things called `libmpcdec` from the same author:

* http://svn.musepack.net/libmpcdec/trunk/src/
* http://svn.musepack.net/libmpc/trunk/include/mpc/

I'll postpone this for now and add a comment.

jtojnar commented on pkgs/development/libraries/gstreamer/base/default.nix:
```diff
@@ -22,28 +27,55 @@ stdenv.mkDerivation rec {
   # TODO How to pass these to Meson?
-  configureFlags = [
+  configureFlags = lib.optionals stdenv.isDarwin [
     "--enable-x11=${if enableX11 then "yes" else "no"}"
```
These seem to be the equivalent of `gl_winsys` option (guessing based on presence of `dispmanx`):

https://gitlab.freedesktop.org/gstreamer/gst-plugins-base/blob/3e460766e865a240834fddad3bccc787a36b4c8a/m4/gst-gl.m4

nh2 replied:
I'll leave the autoconf stuff unmodified for now.

nh2 committed:
gst_all_1.gst-plugins-bad: Fix gsm dependency

nh2 committed:
gst_all_1.gst-plugins-bad: Fix rtmp dependency

nh2 commented on pkgs/development/libraries/gstreamer/good/default.nix:
```diff
@@ -58,10 +62,27 @@ stdenv.mkDerivation rec {
     libXext # present feature but undeclared in meson_options.txt, see https://gitlab.freedesktop.org/gstreamer/gst-plugins-good/issues/553
```
OK, comments removed; the commit e234932dc703e51a0e1aa3b9c408f12758b12335 is in 1.16.

nh2 commented:
@GrahamcOfBorg build gst_all_1

nh2 commented:
I've backed out the evaluation fixes Darwin out of this PR, as that isn't really related to the upgrade, and will handle those with a separate branch.

If ofborg goes green, this is good to merge from my side.

jtojnar commented:
Dependency changes described in [release notes](https://gstreamer.freedesktop.org/releases/1.16/) do not affect us.

What is up with ffmpeg? And do we care about the rust bindings?

nh2 commented:
> Dependency changes described in [release notes](https://gstreamer.freedesktop.org/releases/1.16/) do not affect us.

Thanks for checking!

> What is up with ffmpeg?

I've updated `gst-libav` (which is the ffmpeg binding, right?) in this PR as one of the many gst packages, and it builds successfully.

> And do we care about the rust bindings?

To my knowledge they aren't packaged in nixpkgs yet.

I'll leave that to somebody else, as I have no experience with packaging Rust for nixpkgs, and care mainly about the update to 1.16 and the improved meson build system I've added in this PR.

jtojnar commented on pkgs/development/libraries/gstreamer/good/default.nix:
```diff
@@ -10,6 +10,10 @@
+, xorg
+, libgudev
+, wavpack
+, jack2
```
Jack adds a dependency on Qt through ffado :crying_cat_face:

nh2 replied:
OK, made it optional.

nh2 commented:
@jtojnar Good to go?

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-19-09/2875/24

nh2 commented:
@GrahamcOfBorg build gst_all_1

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/what-are-your-goals-for-19-09/2875/29

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-weekly-11-nixery-nixfmt-and-cachix-releases-nixcon-2019-tickets-a-job-and-first-impressions-post/3591/1
